### PR TITLE
opt_plugin: add result uploader plugin

### PR DESCRIPTION
**Pull request by dmonakhov**
It would be nice to have convinient way to upload results to centralized server.
This plugin superb resultsdb which support external log URL.
Proposed implementation is very simple, but it does it's job.

Changes since V1: in request to @apahim 's comments
 - default rsync upload script  force non interactive mode

### CODE REVIEW

File: optional_plugins/result_upload/avocado_result_upload/__init__.py [hunk 1]
```diff
@@ -0,0 +1,113 @@
...
+
+        msg = 'result-upload options'
+        parser = run_subcommand_parser.add_argument_group(msg)
+        parser.add_argument('--result-upload-url',
+                            dest='result_upload_url', default=None,
+                            help='Specify the result upload url')
+
+        def_ssh = 'ssh -o stricthostkeychecking=no -o userknownhostsfile=/dev/null'
+        def_ssh = def_ssh + ' -o batchmode=yes -o passwordauthentication=no'
+        def_upload_cmd = 'rsync -arz -e \'%s \'' % def_ssh
```
Comment (apahim) [inline optional_plugins/result_upload/avocado_result_upload/__init__.py]: Ok, this improves the situation a lot. Just a last adjustment, let's handle the `rsync` command availability. We have an utils for that:

```python
from avocado.utils import path as utils_path
...

    try:
        rsync_bin = utils_path.find_command('rsync')
    except CmdNotFoundError:
        #handle missing command

...
```

File: optional_plugins/result_upload/avocado_result_upload/__init__.py [hunk 2]
```diff
@@ -0,0 +1,113 @@
...
+
+"""
+Avocado Plugin to propagate Job results to remote host
+"""
+
+import os
+import sys
+import time
+
+import resultsdb_api
```
Comment (apahim) [inline optional_plugins/result_upload/avocado_result_upload/__init__.py]: please remove this unused import

File: optional_plugins/result_upload/avocado_result_upload/__init__.py [hunk 3]
```diff
@@ -0,0 +1,113 @@
...
+from six import iteritems
```
Comment (apahim) [inline optional_plugins/result_upload/avocado_result_upload/__init__.py]: and this

File: optional_plugins/result_upload/avocado_result_upload/__init__.py [hunk 4]
```diff
@@ -0,0 +1,113 @@
...
+# Copyright: Virtuozzo Inc. 2017
+# Authors: Dmitry Monakhov <dmonakhov@openvz.org>
+
+"""
+Avocado Plugin to propagate Job results to remote host
+"""
+
+import os
+import sys
+import time
```
Comment (apahim) [inline optional_plugins/result_upload/avocado_result_upload/__init__.py]: and this

File: optional_plugins/result_upload/avocado_result_upload/__init__.py [hunk 5]
```diff
@@ -0,0 +1,113 @@
...
+#
+# Copyright: Virtuozzo Inc. 2017
+# Authors: Dmitry Monakhov <dmonakhov@openvz.org>
+
+"""
+Avocado Plugin to propagate Job results to remote host
+"""
+
+import os
+import sys
```
Comment (apahim) [inline optional_plugins/result_upload/avocado_result_upload/__init__.py]: wow, and these two :)

File: optional_plugins/result_upload/avocado_result_upload/__init__.py (hunk 1 above)
Comment (dmonakhov) [reply optional_plugins/result_upload/avocado_result_upload/__init__.py]: reasonable. Will do

**Comment by dmonakhov** [comment]
Sorry, I've close this one by accident. PR updated

**Comment by dmonakhov** [comment]
Updated PR.
Changes since V2:
- Add documentation

File: docs/source/optional_plugins/results.rst [hunk 6]
```diff
@@ -30,6 +30,46 @@ specify a custom location via --html . Last but not least
...
+
+To install the ResultUpload plugin from pip, use::
+
+    pip install avocado-framework-plugin-result-upload
+
+Usage::
+
+    avocado run passtest.py --result-upload-url www@avocadologs.example.com:/var/www/html
+
+Avocado logs will available at following URL:
```
Comment (apahim) [inline docs/source/optional_plugins/results.rst]: will **BE** available

File: docs/source/optional_plugins/results.rst [hunk 7]
```diff
@@ -30,6 +30,46 @@ specify a custom location via --html . Last but not least
...
+
+- html (If web server is enabled) ::
+
+    http://avocadologs.example.com/job-2017-04-21T12.54-1cefe11/
+
+Such links may be refered by other plugins, such as ResultsDB
+
+Optionally, you can provide the command which handles upload ::
+
+    avocado run passtest.py --resultsdb-upload-url='s3://bucket' --resultsdb-upload-url='aws s3 cp --recursive'
```
Comment (apahim) [inline docs/source/optional_plugins/results.rst]: This command line needs review.

Comment (apahim) [reply docs/source/optional_plugins/results.rst]: There's no `--resultsdb-upload-url` and `--result-upload-cmd` is missing.

File: docs/source/optional_plugins/results.rst [hunk 8]
```diff
@@ -30,6 +30,46 @@ specify a custom location via --html . Last but not least
...
+You can also set the ResultUpload URL and command using a config file::
+
+    [plugins.resultsdb]
+    url = www@avocadologs.example.com:/var/www/htmlavocado/job-results
+    command='rsync -arzq'
+
+And then run the Avocado command without the explicit cmd options.Notice
+that the command line options will have precedence over the
+configuration file.
+
```
Comment (apahim) [inline docs/source/optional_plugins/results.rst]: Please make sure you tell users what the default command is.

File: optional_plugins/result_upload/avocado_result_upload/__init__.py [hunk 9]
```diff
@@ -0,0 +1,109 @@
...
+            job.log.error("ResultUploader failed msg=%s" % result.stderr)
+
+
+class ResultUploadCLI(CLI):
+
+    """
+    Propagate Job results to Resultsdb
+    """
+
+    name = 'resultsdb'
```
Comment (apahim) [inline optional_plugins/result_upload/avocado_result_upload/__init__.py]: `resultsdb` has nothing to do with this plugin.

File: optional_plugins/result_upload/avocado_result_upload/__init__.py [hunk 10]
```diff
@@ -0,0 +1,109 @@
...
+    description = "Resultsdb options for 'run' subcommand"
```
Comment (apahim) [inline optional_plugins/result_upload/avocado_result_upload/__init__.py]: Ditto

**Comment by dmonakhov** [comment]
What does it means?  Should I update and repoen this PR, or I should start another one? Or you fix it by yourself

**Comment by apahim** [comment]
Best approach is to send a new PR with updated commits and name it as `opt_plugin: add result uploader plugin [v2]`, using the description to point to the previous version and list the changes of the current version when compared to the previous one. Updating the PR, while acceptable, makes the review/merge process harder.

**Comment by apahim** [comment]
oh, in case it was dubious, by "we are close" I meant "we are almost there" :)

**Comment by dmonakhov** [comment]
Ok :)

**Pull request closed without merging**